### ISSUE TITLE
[Discussion] Possible seed init error with Dataloader

### ISSUE THREAD

**Issue by NanoCode012** [ultralytics/yolov5#2761]
## ❔Question
I was reading a [post](https://www.reddit.com/r/MachineLearning/comments/mocpgj/p_using_pytorch_numpy_a_bug_that_plagues/) from /r/MachineLearning when I thought of a previous issue with seed initialization in DDP.

According to that post, there is a possibility that the multiple workers in the Dataloader would use the same seeds to generate identical augmentations. 

I am asking if anyone has heard of this before and whether it was already solved here. If not, I might take a stab at this.

https://github.com/ultralytics/yolov5/blob/6dd1083bbbc5d29643aafef3373853f03a317a92/utils/general.py#L39-L43
https://github.com/ultralytics/yolov5/blob/6dd1083bbbc5d29643aafef3373853f03a317a92/utils/torch_utils.py#L39-L45
https://github.com/ultralytics/yolov5/blob/6dd1083bbbc5d29643aafef3373853f03a317a92/utils/datasets.py#L77-L84

## Possible Solution
The link below lists solutions which others have agreed with, but we may need to test, so that it does not conflict with DDP seed itself.

https://github.com/pytorch/pytorch/issues/5059#issuecomment-404232359
```python
# in train.py, at start of each epoch
np.random.seed()

# in utils/dataset.py
def worker_init_fn(worker_id):                                                          
    np.random.seed(np.random.get_state()[1][0] + worker_id)

# pass function to Dataloader
dataloader = loader(..., worker_init_fn=worker_init_fn)
```

**Comment by glenn-jocher** [comment]
@NanoCode012 that's funny, I saw this as well. If I remember correctly there was an initial bug in ou DDP implementation that was due to the identical seeds, resulting in lower mAP as each rank would apply the same augmentations, but this was resolved soon after.

I think L62 here in train.py should be working well by initializing each rank with it's own separate seed, though I'm not 100% clear on the relationship between DDP rank and dataloader workers.
https://github.com/ultralytics/yolov5/blob/9029759cb3b39de724f148f0c9eee8c70e0ffdc4/train.py#L59-L62

Of course that's just for DDP, for single-GPU there's only one seed init, though we've done comparisons in the past between single and multi-GPU and observed similar mAPs, so I'd be surprised if there existed this type of error in our setup.

**Comment by NanoCode012** [comment]
How about for single GPU runs though? Could there be identical data augmentations silently happening of which we do not know of or have this been tested already?

Edit: I will set two runs to verify this

**Comment by glenn-jocher** [comment]
> How about for single GPU runs though? Could there be identical data augmentations silently happening of which we do not know of or have this been tested already?
> 
> Edit: I will set two runs to verify this

Yes sure, if we had a problem that was common to both single and multi GPU then that might be a reason for seeing the same results between the two, i.e. like a BGR instead of RGB dataloader.

That would actually be great if we had a bug, it would mean there was room to improve the results!

**Comment by NanoCode012** [comment]
I got the results a few days ago but forgot to post it.

It turns out the default settings was 0.1-0.2% better in both Val map0.5 and map0.5.0.95 than the solution posted above. It could be random.

One conjecture I had was that this bug is more likely to occur when new dataloaders workers are created every epoch, therefore setting the correct seed is vital. 

However, due to the Infinite Dataloader, we reuse previous worker, don’t need to recreate them, and may have killed this “bug?”. But, I don’t recall the dataloader increasing map.

I’ll keep this Issue open in case anyone wants to test further or has any new ideas.

**Comment by glenn-jocher** [comment]
@NanoCode012 ok thanks! That's an interesting idea about the infinite dataloader, but I think the mAP stayed the same before and after it's introduction, so it may not have an effect.
